// @carbon/react/lib/components/FormLabel/FormLabel.js
/**
 * Copyright IBM Corp. 2016, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _rollupPluginBabelHelpers = require('../../_virtual/_rollupPluginBabelHelpers.js');
var PropTypes = require('prop-types');
var React = require('react');
var cx = require('classnames');
var usePrefix = require('../../internal/usePrefix.js');
require('../Text/index.js');
var Text = require('../Text/Text.js');

function FormLabel({
  className: customClassName,
  children,
  id,
  ...rest
}) {
  const prefix = usePrefix.usePrefix();
  const className = cx(`${prefix}--label`, `${prefix}--label--no-margin`, customClassName);
  return /*#__PURE__*/React.createElement(Text.Text, _rollupPluginBabelHelpers.extends({
    as: "label",
    htmlFor: id,
    className: className
  }, rest), children);
}
FormLabel.propTypes = {
  /**
   * Specify the content of the form label
   */
  children: PropTypes.node,
  /**
   * Provide a custom className to be applied to the containing <label> node
   */
  className: PropTypes.string,
  /**
   * Provide a unique id for the given <FormLabel>
   */
  id: PropTypes.string
};

exports.default = FormLabel;
